Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,23 +20,40 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return totalAmpHours_;
 }
-
 bool BatteryStateOfChargeService::isCharging() const
 {
+    if(current_ < 0)
+        return true;
     return false;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    QTime timeToCompletion_(0, 0, 0, 0);
+    if(current_ < 0) //charging
+        return (timeToCompletion_.addMSecs(qAbs(totalAmpHoursUsed()/current_)*HOUR_TO_MILLISECOND));
+    else    //depleting
+        return (timeToCompletion_.addMSecs(((BATTERY_AMP_HOUR_CAPACITY - totalAmpHoursUsed())/current_)*HOUR_TO_MILLISECOND));
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // This is where you can update your variables
-    // Hint: There are many different ways that the totalAmpHoursUsed can be updated
-    // i.e: Taking a running average of your data values, using most recent data points, etc.
+    time_ = batteryData.time;
+    current_ = batteryData.current;
+    //time difference is 0 for the first and then previousTime_ is the previous for the next
+    double timeDifference;
+    timeDifference = (double) (previousTime_.msecsTo(time_)*MILLISECOND_TO_HOUR_CONVERSION);
+    double currentAverage;
+    currentAverage = (previousCurrent_ + current_)/2;
+    previousTime_ = batteryData.time;
+    previousCurrent_ = batteryData.current;
+    additionalAmpHours_ += (currentAverage*timeDifference);
+
+    totalAmpHours_ = (BATTERY_AMP_HOUR_CAPACITY*(initialStateOfChargePercent_/100) + additionalAmpHours_);
+    qDebug() << currentAverage << endl;
+    qDebug() << timeDifference << endl;
```

Comment:
Also, iirc I think qDebug already places a new line for you, but I could be wrong on that.